### ISSUE TITLE
Suggest model.eval() in torch.no_grad (and vice versa)

### ISSUE THREAD

**Issue by HaleTom** [pytorch/pytorch#19160]
## 📚 Documentation

`model.eval()`  and `with torch.no_grad` are both commonly used in evaluating a model.

[Confusion exists about whether setting `model.eval()` also sets no_grad](https://discuss.pytorch.org/t/does-model-eval-with-torch-set-grad-enabled-is-train-have-the-same-effect-for-grad-history/17183/5?u=ataraxy)

Would you accept a PR which suggests usage of the other, with words like:

* If evaluating a model's performance, using Module.eval() may also be useful.
* If evaluating a model's performance, using autograd.no_grad may also be useful.

**Comment by zhangguanheng66** [comment]
Need more discussions before making a decision. I see this is a good idea but it will affect many current PyTorch users.

**Comment by SsnL** [comment]
I think the doc change is fine.

**Comment by soumith** [comment]
yea, let's do the doc change. @HaleTom would accept a  PR for this.

**Comment by HaleTom** [comment]
@soumith  I'll get on to it then.


---------

Related:  For convenience, would you consider allowing something like:

```
Module.eval(grad=False)
Module.train(grad=True)
```

To turn off and on gradient tracking, respectively?

(Of course, the default values of `grad=` should be backward compatible.)

Shall I raise a separate issue for this? I'd also be happy to look into it.

**Comment by apaszke** [comment]
No, I think gradient mode is completely independent of eval/train mode of modules. We should point that in the docs, but I don't think an argument like that is a good idae.

**Comment by HaleTom** [comment]
Thanks @apaszke, good point.

**Comment by Yura52** [comment]
To me, it makes sense to include a shortcut that combines both `model.eval()` and `torch.no_grad()`, because scenarios when this combination is needed happen very often.

P.S. Shameless plug: in the library I developed there is one: [`zero.training.Eval`](https://yura52.github.io/zero/reference/training.html#eval). However, I still think it is worth adding something similar to PyTorch itself.

**Comment by HaleTom** [comment]
Thanks for the bump on this!  I'll get something submitted in the next week or so.

**Comment by SsnL** [comment]
Just a thought, what if we also make `model.eval()` a context manager? So people can do things like 
```py
with model.eval(), torch.no_grad():
  val_acc = val(model)
```

**Comment by Yura52** [comment]
> Just a thought, what if we also make `model.eval()` a context manager?

The only difference it makes is that one can omit `model.train()`, because it becomes the "default" mode if `model.eval()` is **always** used as a context-manager, right? As for me, I will anyway use `.train()` and `.eval()` everywhere just before using the model, because I cannot know if some other code changes (or not) my model's mode. So, with my current experience, I don't see how I can benefit from the `model.eval` as a context-manager.

As for [my suggestion](https://github.com/pytorch/pytorch/issues/19160#issuecomment-690961244) above, the only point was to make the "eval-and-no-grad" combination less verbose and easier to use correctly.

**Comment by SsnL** [comment]
The exact reason for my proposal above is that it is less verbose than the current way... Combining context managers is really easy. I agree with @apaszke that we should have independent options for independent settings. So making `.eval` a ctx manager seems both respecting that principle and make users' lives easier.
